### DISCUSSION TITLE
Read & Write in UC tables with polars

### DISCUSSION THREAD

**Discussion by oliverangelil** [unitycatalog/unitycatalog#409] [Ideas]
**Is your feature request related to a problem? Please describe.**

99.9% of the tables we use at the company I work for, don't require spark since they are not terabytes in size.  

**Describe the solution you would like**

[Polars ](https://github.com/pola-rs/polars) has impressive performance benchmarks, and is designed to work with dataframes. 
Would be great to be able to read/write UC delta tables using the UC pathname (not the table UID)

**Comment by VillePuuska** [comment]
Seconded. Would love to be able to get tables from UC directly to Polars dataframes and write dataframes directly to UC tables. I've actually built a [toy project](https://github.com/VillePuuska/Local-Lakehouse) to do exactly this when everything is local, i.e. locally running UC with no auth and tables stored locally.

I think once the Python SDK gets added (see #286 and #309) work on building this capability can start, whether it's included in UC, in Polars, or built as a completely separate package.

**Comment by francisco-ltech** [comment]
@oliverangelil By UC pathname do you mean the namespace? e.g.: catalog.schema.table? I think this is coming soon if I'm not mistaken.

Not sure if it helps, but in the meantime, you can still create your table metadata in UC via bin/uc table --full_name unity.default.data --storage_location "path/to/table" ...

Then, have polars to write the data to the location specified in the previous command.

**Comment by oliverangelil** [comment]
@francisco-ltech yes I mean the namespace (catalog.schema.table). Just like one can read a table with spark using `spark.table(catalog.schema.table)`

**Comment by MrPowers** [comment]
Thanks for opening this.  I converted it to a discussion to allow for better brainstorming.

**Comment by oliverangelil** [comment]
Is this OSS version of UC the version that is used in Databricks? I.e.  if functionality to read/write delta tables with polars is added, what does that mean for users of Databricks?

**Reply by francisco-ltech** [reply]
I believe this is just the baseline of what you get in Databricks, I've seen a few different behaviours that makes me think is still a few steps behind of the cloud version.

**Comment by oliverangelil** [comment]
How about this basic tentative solution - of course there is a cost associated with the conversion: spark -> pandas -> polars; and then back polars -> pandas -> spark. However at least one gets the processing benefits of polars...

```
import polars as pl

def pl_from_uc(uc_path):
    df = spark.table(uc_path)
    pl_df = pl.DataFrame(df.toPandas())
    return pl_df

def pl_to_uc(uc_path, pl_df):
    df = spark.createDataFrame(pl_df.to_pandas())
    df.write.saveAsTable(uc_path)

    
pl_df = pl_from_uc(uc_path="sf_bronze.frlomesweb1._tag_clean")
type(pl_df)  # polars.dataframe.frame.DataFrame
```

**Reply by VillePuuska** [reply]
Reading a Delta table in UC to a Polars dataframe can also be done with DuckDB. Just attach a DuckDB connection `conn` to UC and run

    df = conn.sql("FROM test_catalog.test_schema.test_table").pl()

and `df` will be a Polars dataframe.

No write support for Delta in DuckDB yet though afaik so can't write to UC tables either.

**Comment by hackermandh** [comment]
This, plus the ability to keep using Data Lineage, would be pretty amazing :D

**Comment by francescoalongi** [comment]
This would be amazing. Currently there it seems to be no way to read and write UC managed tables using directly polars. Are there any updates?
